File: questions/127.WordLadder.js
/**
 * ladderLength - BFS
 *
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
const ladderLength = function (beginWord, endWord, wordList) {
  if (!wordList.includes(endWord)) return 0

  const neighbors = {}

  wordList.push(beginWord)

  for (const word of wordList) {
    for (let i = 0; i < word.length; i++) {
      const pattern = word.slice(0, i) + '*' + word.slice(i + 1)
      neighbors[pattern] ||= []
      neighbors[pattern].push(word)
    }
  }

  const visit = new Set()
  visit.add([beginWord])

  const queue = []
  queue.push(beginWord)
  let result = 1

  while (queue.length) {
    const length = queue.length
    for (let i = 0; i < length; i++) {
      const word = queue.shift()
      if (word === endWord) return result

      for (let j = 0; j < word.length; j++) {
        const pattern = word.slice(0, j) + '*' + word.slice(j + 1)
        for (const neighbor of neighbors[pattern]) {
          if (!visit.has(neighbor)) {
            visit.add(neighbor)
            queue.push(neighbor)
          }
        }
      }
    }
    result++
  }

  return 0
}
console.log(
  ladderLength('hit', 'cog', ['hot', 'dot', 'dog', 'lot', 'log', 'cog']),
)

File: leetcode/127.WordLadder.go
package leetcode

func ladderLength(beginWord string, endWord string, wordList []string) int {
	contain := false
	for _, w := range wordList {
		if w == endWord {
			contain = true
		}
	}
	if !contain {
		return 0
	}

	neighbors := make(map[string][]string)

	wordList = append(wordList, beginWord)

	for _, w := range wordList {
		for i := 0; i < len(w); i++ {
			pattern := w[0:i] + "*" + w[i+1:]
			if _, ok := neighbors[pattern]; !ok {
				neighbors[pattern] = []string{}
			}
			neighbors[pattern] = append(neighbors[pattern], w)
		}
	}

	visit := make(map[string]struct{})
	visit[beginWord] = struct{}{}

	var queue []string
	queue = append(queue, beginWord)
	result := 1

	for len(queue) > 0 {
		length := len(queue)
		for i := 0; i < length; i++ {
			word := queue[0]
			queue = queue[1:]

			if word == endWord {
				return result
			}

			for j := 0; j < len(word); j++ {
				pattern := word[0:j] + "*" + word[j+1:]
				for _, neighbor := range neighbors[pattern] {
					if _, ok := visit[neighbor]; !ok {
						visit[neighbor] = struct{}{}
						queue = append(queue, neighbor)
					}
				}
			}
		}
		result++
	}

	return 0
}
